### PR TITLE
Pass all unused arguments to tag helper

### PR DESCRIPTION
**Pull request by calebhearth**
Combined with merging the :classes string/array, and merging :aria, and
:data hashes in the views, this will allow for arbitrary attributes to be passed to
the outermost tag helper, but must be done explicitly for each component
by passing wrapper_options to the tag helper. This avoids the issue of
needing a new attribute and having to add a new kwarg to the component
signature, attr_reader in the component, and manually pass that argument
into the tag helper in the view.

Any argument that is passed to super in **kwargs or explicitly (in the
case of defaulting arguments) will be stored in the wrapper_options
attribute for the component.

The exception is :classes, which will be renamed as :class.

Along the way, extract the VARIANT_* pattern of passing some "named"
variant to the component and adding classes for that component.
Provided that VARIANT_DEFAULT, VARIANT_MAPPINGS, and VARIANT_OPTIONS are
defined on a component, the appropriate classes will be looked up and
merged into the :class argument in wrapper_options.

Co-authored-by: Tyson Gach <tyson@tysongach.com>

### CODE REVIEW

File: app/components/anchor/component.rb [hunk 1]
```diff
@@ -7,16 +7,38 @@ def self.generate_id(base_name: name.demodulize.underscore.dasherize)
...
-    def initialize(classes: nil, data: {}, id: nil, **kwargs)
-      @classes = classes
-      @data = data
-      @id = id
+    def initialize(**kwargs)
+      classes = add_variant_classes_if_available(
+        classes: kwargs.delete(:classes) || [],
+        variant: kwargs[:variant]
+      )
+      @wrapper_options = kwargs.merge(class: classes)
```
Comment (tysongach) [inline app/components/anchor/component.rb:15]: `wrapper_options` didn’t read very clearly to me, I wonder if something more direct like `component_options` could work?

```suggestion
      @component_options = kwargs.merge(class: classes)
```

**Review by tysongach — COMMENTED** [review]
I like this! I’d also like to hear from others, @Simonpedro, @dgalarza, @tute…

You mentioned to me that this breaks some existing components and I see we have some failing tests - do you have a path forward? What would need to change?

**Comment by dgalarza** [comment]
> I like this! I’d also like to hear from others, @Simonpedro, @dgalarza, @tute…
> 
> You mentioned to me that this breaks some existing components and I see we have some failing tests - do you have a path forward? What would need to change?

I dig it myself.

**Comment by Simonpedro** [comment]
I recently had the issue of trying to pass an attribute to the button component that is not explicitly used so I like this!

**Review by Simonpedro — APPROVED** [review]
👏

File: bin/rspec
Comment (masonforest) [inline bin/rspec:1]: I've been getting these too 😕

**Review by masonforest — APPROVED** [review]
Looks good! 🚢

**Pull request merged**